### DISCUSSION TITLE
Need help applying an effect

### DISCUSSION THREAD

**Discussion by oussamabonnor1** [iTwin/itwinjs-core#749] [Q&A] [answered]
Based on the docs, the `addScreenSpaceEffect` is the function to call when applying something like the lens distortion. 
It's called from an instance of the current `ViewPort` which can be acquired (as far as I understand) from either a `IModelApp.viewManager.selectedView` or (in case it was undefined, which in my case it always was) using the `onViewOpen.addOnce()` callback.

I copied the `LensDistortionEffect` from the effect examples in the [frontend-devtools](https://github.com/imodeljs/imodeljs/tree/master/core/frontend-devtools/src/effects) and tried using it but it did not work, I tested other examples like the `FlipImage` one but the same results were met.

I even tried replacing the Toolbar functions with an effect appliance (in case the callback was wrong) using the `IModelApp.tools.run()` but still got nothing.

I was unable to find any examples of applying a post-processing effect in the docs.

I would love to get some help on how to apply an effect.

**Comment by pmconne** [comment, marked as answer]
Hi @oussamabonnor1. Can you show your code (minus that which was copied from frontend-devtools)?

**Reply by oussamabonnor1** [reply]
Hi @pmconne, please bear in mind that my experience with both IModel.js and React is limited so the following implementation might be far from decent.

I am using the basic-viewport-app found in the interactive-app folder of the [imodeljs-samples](https://github.com/imodeljs/imodeljs-samples) repo.

In the [src/frontend/components/App.tsx](https://github.com/imodeljs/imodeljs-samples/blob/master/interactive-app/basic-viewport-app/src/frontend/components/App.tsx) I added the following function:

```
public registerEffect(name: string, alterColor: string): void {
    const fragment = `
      vec4 effectMain() {
        vec4 color = TEXTURE(u_diffuse, v_texCoord);
        ${alterColor}
        return color;
      }`;

    const builder = IModelApp.renderSystem.createScreenSpaceEffectBuilder({
      name,
      textureCoordFromPosition: true,
      source: {
        vertex: "void effectMain(vec4 pos) { v_texCoord = textureCoordFromPosition(pos); }",
        fragment,
      },
    })!;

    builder.shouldApply = () => disabledEffectName !== name;
    builder.addVarying("v_texCoord", VaryingType.Vec2);

    builder.finish();
  }
```
I called the function here:
```
  public componentDidMount() {
    IModelApp.viewManager.onViewOpen.addOnce((vp) => {
      this.registerEffect("BGC", "color.rgb = color.b > 0.0 ? vec3(0, 1, 0) : vec3(0, 1, 1);");
      let effects = vp.screenSpaceEffects;
      console.log("effects", effects)
    });

    // Initialize authorization state, and add listener to changes
    BasicViewportApp.oidcClient.onUserStateChanged.addListener(this._onUserStateChanged);
  }
```

This was heavily inspired from the full-stack-tests in the [IModeljs repo](https://github.com/imodeljs/imodeljs). 

**I honestly would love to have an example about how a basic effect can be applied following best practice guidelines.**

**Reply by pmconne** [reply]
Thanks @oussamabonnor1. In your `addOnce` callback you're registering your effect, but I don't see you calling `addScreenSpaceEffect` anywhere. Is that missing from your actual code or just omitted from the snippets you provided here?

I should be able to open a PR to add a sample showing how to use screen-space effects to the [sample showcase](https://www.itwinjs.org/sample-showcase/) today.

**Reply by oussamabonnor1** [reply]
Thank you for pointing that out @pmconne, I was indeed missing that call which made using the "BGC" effect possible.
Please proceed with the screen-space effect sample, that would be of huge help to us.
I will meanwhile continue experimenting with the Lens Distortion effect since I now managed to apply the previous effect.

**Reply by oussamabonnor1** [reply]
Hi @pmconne, I couldn't help but notice that the PR for the screen space effect samples was merged.
I was wondering where I can  find it in the samples? 
Considering the [sampleManifest](https://github.com/imodeljs/frontend-sample-showcase/blob/master/src/sampleManifest.tsx), it should be in the [Viewer Features](https://github.com/imodeljs/frontend-sample-showcase/blob/73a73ae078ccd6e4a549b2f462a1c4fda16158e8/src/sampleManifest.tsx#L78) between `Read Settings` & `Shadow Study` but I cannot seem to be able to find it.

**Reply by pmconne** [reply]
You can pull, build, and run it locally now. I'm not sure when @Josh-Schifter or @williamkbentley plan to update the showcase site.

**Reply by oussamabonnor1** [reply]
Understood, thank you for your quick reply!

**Reply by Josh-Schifter** [reply]
The new sample is deployed here: https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=screen-space-effects-sample&imodel=Villa
